server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/endpoint/EstimateSampleSizes.java

```diff
@@ -1,73 +1,258 @@
+    // OPPORTUNISTIC_BENEFITS as the reason for each audit. The tabulated vote totals
+    // in a ContestResult for an IRV contest will not be used. In the call to ContestCounter
+    // (countAllContests), all persisted CountyContestResults will be accessed from the database,
+    // grouped by contest, and accumulated into a single ContestResult.
+    final List<ContestResult> countedCRs = ContestCounter.countAllContests().stream().peek(cr ->
+        cr.setAuditReason(AuditReason.OPPORTUNISTIC_BENEFITS)).toList();
+
+    // Try to get the DoS Dashboard, which may contain the risk limit for the audit.
+    riskLimit = Persistence.getByID(DoSDashboard.ID, DoSDashboard.class).auditInfo().riskLimit();
+    if (riskLimit == null) {
+      // If the risk limit is not initialized, it is not possible to estimate sample sizes.
+      String msg = "No risk limit set";
+      LOGGER.error(String.format("%s %s.", prefix, msg));
+      throw new IllegalStateException(msg);
     }
 
-    /**
-     * @return STATE authorization is necessary for this endpoint.
-     */
-    public AuthorizationType requiredAuthorization() {
-        return AuthorizationType.STATE;
+    // Iterate over all the contest results, getting relevant data into an EstimateData record.
+    for (ContestResult cr : countedCRs) {
+
+      final ComparisonAudit ca = createAuditOfCorrectType(cr, riskLimit);
+      final List<String> countyNames = ca.getCounties().stream().map(County::name).toList();
+
+      // The total number of ballots cast that have that contest on it. This is equal to the sum
+      // of my_contest_count in the county contest results for the contest.
+      final int contestBallots = CountyContestResultQueries.withContestName(ca.getContestName())
+          .stream().mapToInt(CountyContestResult::contestBallotCount).sum();
+
+      dataRows.add(new EstimateData(countyNames, ca, contestBallots, cr.getBallotCount()));
     }
 
+    // Sort the data according to the name of the county, and then by contest.
+    Collections.sort(dataRows);
+    // dataRows.sort((first, second) -> first.countyName.compareTo(second.countyName()));
+    return String.join("\n", dataRows.stream().map(EstimateData::toString).toList());
+  }
+
+  /**
+   * A record to store the data we need in the report. Stores relevant data (in some cases after
+   * retrieving it from more complex structures) and outputs it as a CSV row (toString()).
+   *
+   * @param countyName       The name of the county.
+   * @param contestName      The name of the contest.
+   * @param contestType      The type, as a string, IRV or plurality.
+   * @param contestBallots   The number of ballots that actually contained the contest.
+   * @param totalBallots     The total universe of ballots for this contest's audit.
+   * @param dilutedMargin    The margin divided by the total Auditable ballots.
+   * @param estimatedSamples The estimated samples to audit.
```

Comment:
What input arguments can be final?

Reply:
Turns out record constructor parameters cannot be final.